File: packages/js/src/providerData.ts
import type { ConditionalPrice, Provider } from './types'

import { utcTimeOfDaySeconds } from './timeOfDay'
import { UnitRegistry, validateUnitEvolution } from './units'

export type DecodedProviderData = Readonly<{
  compatibilityWarnings: string[]
  providers: Provider[]
  registry?: UnitRegistry
}>

type Projection = Readonly<{ supported: boolean; value: unknown }>
type WireExtractor = Record<string, unknown> & {
  api_flavor?: unknown
  mappings: unknown[]
  model_path?: unknown
}
type WireExtractorMapping = Record<string, unknown> & { required?: unknown }

export function decodeProviderData(raw: unknown, compatibilityRegistry: UnitRegistry): DecodedProviderData {
  if (Array.isArray(raw)) return decodeLegacyProviderArray(raw)
  if (isRecord(raw)) return decodeWrappedProviderData(raw, compatibilityRegistry)
  throw invalidData('root must be a wrapped object or provider array')
}

function decodeLegacyProviderArray(raw: unknown[]): DecodedProviderData {
  return { compatibilityWarnings: [], providers: normalizeProviderArray(raw) }
}

function decodeWrappedProviderData(raw: Record<string, unknown>, compatibilityRegistry: UnitRegistry): DecodedProviderData {
  if (!Object.prototype.hasOwnProperty.call(raw, 'units')) throw invalidData('missing units')
  if (!Object.prototype.hasOwnProperty.call(raw, 'providers')) throw invalidData('missing providers')
  if (!Array.isArray(raw.providers)) throw invalidData('providers must be an array')

  const registry = UnitRegistry.fromUntrusted(raw.units)
  validateUnitEvolution(compatibilityRegistry, registry)
  const projector = new ProviderProjector()
  const projectedProviders = raw.providers.map((provider, index) => projector.projectProvider(provider, index))
  return {
    compatibilityWarnings: projector.warnings,
    providers: normalizeProviderArray(projectedProviders),
    registry,
  }
}

class ProviderProjector {
  readonly warnings: string[] = []

  projectProvider(raw: unknown, providerIndex: number): unknown {
    if (!isRecord(raw)) return raw
    const provider = { ...raw }
    const context = providerContext(provider, providerIndex)

    for (const field of ['model_match', 'provider_match'] as const) {
      if (field in provider) {
        const path = `providers[${String(providerIndex)}].${field}`
        const projected = this.projectMatch(provider[field], path)
        if (projected.supported) provider[field] = projected.value
        else {
          this.warn('match', path, context)
          Reflect.deleteProperty(provider, field)
        }
      }
    }

    if (Array.isArray(provider.extractors)) {
      const extractors: unknown[] = []
      for (const [index, extractor] of provider.extractors.entries()) {
        const projected = this.projectExtractor(extractor, `providers[${String(providerIndex)}].extractors[${String(index)}]`, context)
        if (projected !== undefined) extractors.push(projected)
      }
      provider.extractors = extractors
    }

    if (Array.isArray(provider.models)) {
      const models: unknown[] = []
      for (const [index, model] of provider.models.entries()) {
        const projected = this.projectModel(model, providerIndex, index, context)
        if (projected !== undefined) models.push(projected)
      }
      provider.models = models
    }
    return provider
  }

  private projectExtractor(raw: unknown, path: string, context: string): unknown {
    if (!isRecord(raw)) return raw
    const extractor = { ...raw }
    if ('type' in extractor || (!('root' in extractor) && !('mappings' in extractor))) {
      this.warn('extractor', path, context)
      return undefined
    }
    if (!('root' in extractor)) throw invalidData(`${path}.root is required`)
    if (!('mappings' in extractor)) throw invalidData(`${path}.mappings is required`)

    for (const field of ['root', 'model_path'] as const) {
      if (field in extractor) {
        const fieldPath = `${path}.${field}`
        const projected = this.projectExtractPath(extractor[field], fieldPath)
        if (!projected.supported) {
          this.warn('extractor', fieldPath, context)
          return undefined
        }
        extractor[field] = projected.value
      }
    }

    if (Array.isArray(extractor.mappings)) {
      const originalMappingCount = extractor.mappings.length
      const mappings: unknown[] = []
      for (const [index, rawMapping] of extractor.mappings.entries()) {
        if (!isRecord(rawMapping)) {
          mappings.push(rawMapping)
          continue
        }
        const mappingPath = `${path}.mappings[${String(index)}]`
        const mapping = { ...rawMapping }
        if (!('path' in mapping) && !('dest' in mapping)) {
          this.warn('extractor mapping', mappingPath, context)
          continue
        }
        if ('path' in mapping) {
          const pathPath = `${mappingPath}.path`
          const projected = this.projectExtractPath(mapping.path, pathPath)
          if (!projected.supported) {
            this.warn('extractor mapping', pathPath, context)
            continue
          }
          mapping.path = projected.value
        }
        mappings.push(mapping)
      }
      if (originalMappingCount > 0 && mappings.length === 0) return undefined
      extractor.mappings = mappings
    }
    return extractor
  }

  private projectExtractPath(raw: unknown, path: string): Projection {
    if (typeof raw === 'string') return { supported: true, value: raw }
    if (!Array.isArray(raw)) {
      if (isRecord(raw)) return { supported: false, value: raw }
      throw invalidData(`${path} must be a string or array`)
    }

    const steps: unknown[] = []
    for (const [index, step] of raw.entries()) {
      if (typeof step === 'string') {
        steps.push(step)
        continue
      }
      if (!isRecord(step)) throw invalidData(`${path}[${String(index)}] must be a string or object`)
      if (step.type !== 'array-match') return { supported: false, value: raw }
      const arrayMatch = { ...step }
      if (typeof arrayMatch.field !== 'string') throw invalidData(`${path}[${String(index)}].field must be a string`)
      if (!('match' in arrayMatch)) throw invalidData(`${path}[${String(index)}].match is required`)
      const projected = this.projectMatch(arrayMatch.match, `${path}[${String(index)}].match`)
      if (!projected.supported) return { supported: false, value: raw }
      arrayMatch.match = projected.value
      steps.push(arrayMatch)
    }
    return { supported: true, value: steps }
  }

  private projectMatch(raw: unknown, path: string): Projection {
    if (!isRecord(raw)) throw invalidData(`${path} match must be an object`)
    const discriminator = matchDiscriminators.find((key) => key in raw)
    if (discriminator === undefined) return { supported: false, value: raw }
    if ((discriminator === 'or' || discriminator === 'and') && Array.isArray(raw[discriminator])) {
      const children: unknown[] = []
      for (const [index, child] of raw[discriminator].entries()) {
        const projected = this.projectMatch(child, `${path}.${discriminator}[${String(index)}]`)
        if (!projected.supported) return { supported: false, value: raw }
        children.push(projected.value)
      }
      return { supported: true, value: { ...raw, [discriminator]: children } }
    }
    return { supported: true, value: raw }
  }

  private projectModel(raw: unknown, providerIndex: number, modelIndex: number, providerContextValue: string): unknown {
    if (!isRecord(raw)) return raw
    const model = { ...raw }
    const path = `providers[${String(providerIndex)}].models[${String(modelIndex)}]`
    const context = `${providerContextValue}, ${modelContext(model, modelIndex)}`
    if ('match' in model) {
      const projected = this.projectMatch(model.match, `${path}.match`)
      if (!projected.supported) {
        this.warn('match', `${path}.match`, context)
        return undefined
      }
      model.match = projected.value
    }
    if ('prices' in model) {
      const rawPrices = model.prices
      const projectedPrices = this.projectPrices(rawPrices, `${path}.prices`, context)
      if (
        isRecord(rawPrices) &&
        Object.keys(rawPrices).length > 0 &&
        isRecord(projectedPrices) &&
        Object.keys(projectedPrices).length === 0
      ) {
        return undefined
      }
      if (Array.isArray(rawPrices) && rawPrices.length > 0 && Array.isArray(projectedPrices) && projectedPrices.length === 0) {
        return undefined
      }
      model.prices = projectedPrices
    }
    return model
  }

  private projectPriceMap(raw: Record<string, unknown>, path: string, context: string): Record<string, unknown> {
    const prices: Record<string, unknown> = {}
    for (const [priceKey, value] of Object.entries(raw)) {
      if (isRecord(value) && !('base' in value) && !('tiers' in value)) {
        this.warn('price', `${path}.${priceKey}`, context)
      } else prices[priceKey] = value
    }
    return prices
  }

  private projectPrices(raw: unknown, path: string, context: string): unknown {
    if (isRecord(raw)) return this.projectPriceMap(raw, path, context)
    if (!Array.isArray(raw)) return raw

    const prices: unknown[] = []
    for (const [index, rawPrice] of raw.entries()) {
      if (!isRecord(rawPrice)) {
        prices.push(rawPrice)
        continue
      }
      const pricePath = `${path}[${String(index)}]`
      if (!('prices' in rawPrice)) {
        this.warn('price', pricePath, context)
        continue
      }
      const conditionalPrice = { ...rawPrice }
      if ('constraint' in conditionalPrice) {
        const projected = projectConstraint(conditionalPrice.constraint)
        if (!projected.supported) {
          this.warn('constraint', `${pricePath}.constraint`, context)
          continue
        }
        conditionalPrice.constraint = projected.value
      }
      if (isRecord(conditionalPrice.prices)) {
        const rawPriceCount = Object.keys(conditionalPrice.prices).length
        const projectedPrices = this.projectPriceMap(conditionalPrice.prices, `${pricePath}.prices`, context)
        if (rawPriceCount > 0 && Object.keys(projectedPrices).length === 0) continue
        conditionalPrice.prices = projectedPrices
      }
      prices.push(conditionalPrice)
    }
    return prices
  }

  private warn(capability: string, path: string, context: string): void {
    this.warnings.push(`Unsupported ${capability} variant at ${path} for ${context}; upgrade genai-prices for full support`)
  }
}

const matchDiscriminators = ['or', 'and', 'equals', 'starts_with', 'ends_with', 'contains', 'regex'] as const

function hasAny(record: Record<string, unknown>, keys: readonly string[]): boolean {
  return keys.some((key) => key in record)
}

function invalidData(message: string): Error {
  return new Error(`genai-prices: invalid data: ${message}`)
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function modelContext(model: Record<string, unknown>, index: number): string {
  return typeof model.id === 'string' ? `model '${model.id}'` : `model index ${String(index)}`
}

function normalizeConditionalPrice(raw: unknown, providerId: string, modelId: string, path: string): ConditionalPrice {
  if (!isRecord(raw)) throw invalidData(`${path} must be an object for provider '${providerId}' model '${modelId}'`)
  if (!isRecord(raw.prices)) {
    throw invalidData(`${path}.prices must be an object for provider '${providerId}' model '${modelId}'`)
  }
  const constraint = raw.constraint
  const prices = raw.prices as ConditionalPrice['prices']
  if (constraint === undefined) return { prices }
  if (!isRecord(constraint)) throw invalidConstraintError(constraint, providerId, modelId, path)
  if (constraint.type !== undefined) {
    if (
      (constraint.type === 'start_date' && hasExactKeys(constraint, ['start_date', 'type']) && isValidStartDate(constraint.start_date)) ||
      (constraint.type === 'time_of_date' &&
        hasExactKeys(constraint, ['end_time', 'start_time', 'type']) &&
        isValidTimeOfDay(constraint.start_time) &&
        isValidTimeOfDay(constraint.end_time))
    ) {
      return raw as unknown as ConditionalPrice
    }
    throw invalidConstraintError(constraint, providerId, modelId, path)
  }
  if (hasExactKeys(constraint, ['start_date']) && isValidStartDate(constraint.start_date)) {
    return { constraint: { start_date: constraint.start_date, type: 'start_date' }, prices }
  }
  if (
    hasExactKeys(constraint, ['end_time', 'start_time']) &&
    isValidTimeOfDay(constraint.start_time) &&
    isValidTimeOfDay(constraint.end_time)
  ) {
    return {
      constraint: { end_time: constraint.end_time, start_time: constraint.start_time, type: 'time_of_date' },
      prices,
    }
  }
  throw invalidConstraintError(constraint, providerId, modelId, path)
}

function normalizeExtractor(raw: unknown, path: string): unknown {
  if (!isRecord(raw)) throw invalidData(`${path} must be an object`)
  if (!Array.isArray(raw.mappings)) throw invalidData(`${path}.mappings must be an array`)
  const extractor: WireExtractor = { ...raw, mappings: raw.mappings }
  return {
    ...extractor,
    api_flavor: extractor.api_flavor ?? 'default',
    mappings: extractor.mappings.map((mapping, index) => {
      if (!isRecord(mapping)) throw invalidData(`${path}.mappings[${String(index)}] must be an object`)
      if (!('path' in mapping) || !('dest' in mapping)) {
        throw invalidData(`${path}.mappings[${String(index)}] must include path and dest`)
      }
      const wireMapping: WireExtractorMapping = mapping
      return { ...wireMapping, required: wireMapping.required ?? true }
    }),
    model_path: extractor.model_path ?? 'model',
  }
}

function normalizeModel(raw: unknown, providerId: string, providerIndex: number, modelIndex: number): unknown {
  const path = `providers[${String(providerIndex)}].models[${String(modelIndex)}]`
  if (!isRecord(raw)) throw invalidData(`${path} must be an object`)
  if (!('match' in raw)) throw invalidData(`${path}.match is required`)
  if (!isRecord(raw.prices) && !Array.isArray(raw.prices)) {
    throw invalidData(`${path}.prices must be an object or array`)
  }
  const modelId = typeof raw.id === 'string' ? raw.id : String(modelIndex)
  return {
    ...raw,
    prices: Array.isArray(raw.prices)
      ? raw.prices.map((price, priceIndex) =>
          normalizeConditionalPrice(price, providerId, modelId, `${path}.prices[${String(priceIndex)}]`)
        )
      : raw.prices,
  }
}

function normalizeProviderArray(raw: unknown[]): Provider[] {
  return raw.map((rawProvider, providerIndex) => {
    const path = `providers[${String(providerIndex)}]`
    if (!isRecord(rawProvider)) throw invalidData(`${path} must be an object`)
    if (!Array.isArray(rawProvider.models)) throw invalidData(`${path}.models must be an array`)
    const providerId = typeof rawProvider.id === 'string' ? rawProvider.id : String(providerIndex)
    if (rawProvider.extractors !== undefined && !Array.isArray(rawProvider.extractors)) {
      throw invalidData(`${path}.extractors must be an array`)
    }
    return {
      ...rawProvider,
      ...(rawProvider.extractors === undefined
        ? {}
        : {
            extractors: rawProvider.extractors.map((extractor, index) =>
              normalizeExtractor(extractor, `${path}.extractors[${String(index)}]`)
            ),
          }),
      models: rawProvider.models.map((model, index) => normalizeModel(model, providerId, providerIndex, index)),
    } as unknown as Provider
  })
}

function projectConstraint(raw: unknown): Projection {
  if (!isRecord(raw)) return { supported: true, value: raw }
  if ('type' in raw) {
    if (raw.type === 'start_date') {
      if (!('start_date' in raw)) return { supported: true, value: raw }
      if (hasAny(raw, ['end_time', 'start_time'])) return { supported: true, value: raw }
      return { supported: true, value: { start_date: raw.start_date, type: raw.type } }
    }
    if (raw.type === 'time_of_date') {
      if (!('start_time' in raw) || !('end_time' in raw)) return { supported: true, value: raw }
      if ('start_date' in raw) return { supported: true, value: raw }
      return { supported: true, value: { end_time: raw.end_time, start_time: raw.start_time, type: raw.type } }
    }
    return { supported: false, value: raw }
  }

  const hasStartDate = 'start_date' in raw
  const hasTime = hasAny(raw, ['start_time', 'end_time'])
  if (hasStartDate && hasTime) return { supported: true, value: raw }
  if (hasStartDate) return { supported: true, value: { start_date: raw.start_date } }
  if (hasTime) {
    if (!('start_time' in raw) || !('end_time' in raw)) return { supported: true, value: raw }
    return { supported: true, value: { end_time: raw.end_time, start_time: raw.start_time } }
  }
  return { supported: false, value: raw }
}

function providerContext(provider: Record<string, unknown>, index: number): string {
  return typeof provider.id === 'string' ? `provider '${provider.id}'` : `provider index ${String(index)}`
}

function invalidConstraintError(constraint: unknown, providerId: string, modelId: string, path: string): Error {
  return invalidData(
    `${path} expected a start-date or time-of-day price constraint for provider '${providerId}' model '${modelId}', got: ${JSON.stringify(constraint)}`
  )
}

function isValidStartDate(value: unknown): value is string {
  if (typeof value !== 'string' || !/^\d{4}-\d{2}-\d{2}$/.test(value)) return false
  const parsed = new Date(`${value}T00:00:00Z`)
  return !Number.isNaN(parsed.getTime()) && parsed.getUTCFullYear() >= 1 && parsed.toISOString().startsWith(value)
}

function hasExactKeys(value: Record<string, unknown>, keys: string[]): boolean {
  const actual = Object.keys(value)
  return actual.length === keys.length && actual.every((key) => keys.includes(key))
}

function isValidTimeOfDay(value: unknown): value is string {
  if (typeof value !== 'string') return false
  try {
    utcTimeOfDaySeconds(value)
    return true
  } catch {
    return false
  }
}

File: packages/python/genai_prices/provider_data.py
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, cast

from .types import Provider, _providers_from_raw  # pyright: ignore[reportPrivateUsage]
from .units import UnitRegistry, _validate_unit_evolution  # pyright: ignore[reportPrivateUsage]


@dataclass(frozen=True)
class _DecodedProviderData:
    providers: list[Provider]
    registry: UnitRegistry | None
    compatibility_warnings: tuple[str, ...]


def _decode_provider_data(  # pyright: ignore[reportUnusedFunction]
    raw: object, compatibility_registry: UnitRegistry
) -> _DecodedProviderData:
    if isinstance(raw, list):
        return _decode_legacy_provider_array(cast(list[object], raw))
    if isinstance(raw, Mapping):
        return _decode_wrapped_provider_data(cast(Mapping[str, object], raw), compatibility_registry)
    raise ValueError('Invalid provider data root: expected a wrapped object or provider array')


def _decode_wrapped_provider_data(
    raw: Mapping[str, object], compatibility_registry: UnitRegistry
) -> _DecodedProviderData:
    if 'units' not in raw:
        raise ValueError('Invalid provider data: missing units')
    if 'providers' not in raw:
        raise ValueError('Invalid provider data: missing providers')
    raw_providers = raw['providers']
    if not isinstance(raw_providers, list):
        raise ValueError('Invalid provider data at providers: expected an array')

    registry = UnitRegistry._from_untrusted(raw['units'])  # pyright: ignore[reportPrivateUsage]
    _validate_unit_evolution(compatibility_registry, registry)
    projected_providers, compatibility_warnings = _project_providers(cast(list[object], raw_providers))
    try:
        providers = _providers_from_raw(projected_providers)
    except (AssertionError, ValueError) as exc:
        raise ValueError(f'Invalid provider data at providers: {exc}') from exc
    return _DecodedProviderData(providers, registry, compatibility_warnings)


def _decode_legacy_provider_array(raw: list[object]) -> _DecodedProviderData:
    return _DecodedProviderData(_providers_from_raw(raw), None, ())


class _ProviderProjector:
    def __init__(self) -> None:
        self.warnings: list[str] = []

    def project_provider(self, raw: object, provider_index: int) -> object:
        if not isinstance(raw, Mapping):
            return raw
        provider = dict(cast(Mapping[str, Any], raw))
        context = _provider_context(provider, provider_index)

        for field in ('model_match', 'provider_match'):
            if field in provider:
                path = f'providers[{provider_index}].{field}'
                supported, projected = self._project_match(provider[field], path)
                if supported:
                    provider[field] = projected
                else:
                    self._warn('match', path, context)
                    del provider[field]

        raw_extractors = provider.get('extractors')
        if isinstance(raw_extractors, list):
            extractors: list[object] = []
            for extractor_index, raw_extractor in enumerate(cast(list[object], raw_extractors)):
                path = f'providers[{provider_index}].extractors[{extractor_index}]'
                projected = self._project_extractor(raw_extractor, path, context)
                if projected is not None:
                    extractors.append(projected)
            provider['extractors'] = extractors

        raw_models = provider.get('models')
        if isinstance(raw_models, list):
            models: list[object] = []
            for model_index, raw_model in enumerate(cast(list[object], raw_models)):
                projected = self._project_model(raw_model, provider_index, model_index, context)
                if projected is not None:
                    models.append(projected)
            provider['models'] = models
        return provider

    def _project_extractor(self, raw: object, path: str, context: str) -> object | None:
        if not isinstance(raw, Mapping):
            return raw
        extractor = dict(cast(Mapping[str, Any], raw))
        if 'type' in extractor or not ({'root', 'mappings'} & extractor.keys()):
            self._warn('extractor', path, context)
            return None

        for field in ('root', 'model_path'):
            if field in extractor:
                field_path = f'{path}.{field}'
                supported, projected = self._project_extract_path(extractor[field], field_path)
                if not supported:
                    self._warn('extractor', field_path, context)
                    return None
                extractor[field] = projected

        raw_mappings = extractor.get('mappings')
        if isinstance(raw_mappings, list):
            mappings: list[object] = []
            for mapping_index, raw_mapping in enumerate(cast(list[object], raw_mappings)):
                mapping_path = f'{path}.mappings[{mapping_index}]'
                if isinstance(raw_mapping, Mapping):
                    mapping = dict(cast(Mapping[str, Any], raw_mapping))
                    if 'type' in mapping or not ({'path', 'dest'} & mapping.keys()):
                        self._warn('extractor mapping', mapping_path, context)
                        continue
                    if 'path' in mapping:
                        path_path = f'{mapping_path}.path'
                        supported, projected = self._project_extract_path(mapping['path'], path_path)
                        if not supported:
                            self._warn('extractor mapping', path_path, context)
                            continue
                        mapping['path'] = projected
                    mappings.append(mapping)
                else:
                    mappings.append(raw_mapping)
            extractor['mappings'] = mappings
        return extractor

    def _project_extract_path(self, raw: object, path: str) -> tuple[bool, object]:
        if isinstance(raw, str):
            return True, raw
        if not isinstance(raw, list):
            return not isinstance(raw, Mapping), raw
        projected_steps: list[object] = []
        for index, step in enumerate(cast(list[object], raw)):
            if isinstance(step, str):
                projected_steps.append(step)
                continue
            if not isinstance(step, Mapping):
                projected_steps.append(step)
                continue
            array_match = dict(cast(Mapping[str, object], step))
            if array_match.get('type') != 'array-match':
                return False, cast(object, raw)
            if 'match' in array_match:
                supported, projected = self._project_match(array_match['match'], f'{path}[{index}].match')
                if not supported:
                    return False, cast(object, raw)
                array_match['match'] = projected
            projected_steps.append(array_match)
        return True, projected_steps

    def _project_model(
        self, raw: object, provider_index: int, model_index: int, provider_context: str
    ) -> object | None:
        if not isinstance(raw, Mapping):
            return raw
        model = dict(cast(Mapping[str, Any], raw))
        path = f'providers[{provider_index}].models[{model_index}]'
        context = f'{provider_context}, {_model_context(model, model_index)}'
        if 'match' in model:
            supported, projected = self._project_match(model['match'], f'{path}.match')
            if not supported:
                self._warn('match', f'{path}.match', context)
                return None
            model['match'] = projected
        if 'prices' in model:
            projected_prices = self._project_prices(model['prices'], f'{path}.prices', context)
            if isinstance(model['prices'], list) and model['prices'] and projected_prices == []:
                return None
            model['prices'] = projected_prices
        return model

    def _project_match(self, raw: object, path: str) -> tuple[bool, object]:
        if not isinstance(raw, Mapping):
            raise ValueError(f'Invalid match at {path}: expected an object')
        match = cast(Mapping[str, object], raw)
        if not match:
            return False, cast(object, raw)
        known_discriminators = {'starts_with', 'ends_with', 'contains', 'regex', 'equals', 'or', 'and'}
        discriminators = [key for key in match if key in known_discriminators]
        if not discriminators:
            return False, cast(object, raw)
        if len(discriminators) != 1:
            raise ValueError(f'Invalid match at {path}: expected exactly one recognized discriminator')
        discriminator = discriminators[0]
        projected_match = {discriminator: match[discriminator]}
        projected_match.update((key, value) for key, value in match.items() if key != discriminator)
        if discriminator not in {'or', 'and'} or not isinstance(match[discriminator], list):
            return True, projected_match

        # Nested match entries are not independently decoded here: retaining an
        # incomplete boolean expression could broaden or narrow model matching.
        projected_children: list[object] = []
        for index, child in enumerate(cast(list[object], match[discriminator])):
            supported, projected = self._project_match(child, f'{path}.{discriminator}[{index}]')
            if not supported:
                return False, cast(object, raw)
            projected_children.append(projected)
        projected_match[discriminator] = projected_children
        return True, projected_match

    def _project_prices(self, raw: object, path: str, context: str) -> object:
        if isinstance(raw, Mapping):
            return self._project_price_map(cast(Mapping[str, object], raw), path, context)
        if not isinstance(raw, list):
            raise ValueError(f'Invalid prices at {path}: expected an object or array')

        prices: list[object] = []
        for price_index, raw_price in enumerate(cast(list[object], raw)):
            price_path = f'{path}[{price_index}]'
            if not isinstance(raw_price, Mapping):
                prices.append(raw_price)
                continue
            conditional = dict(cast(Mapping[str, Any], raw_price))
            if 'prices' not in conditional:
                self._warn('price', price_path, context)
                continue
            constraint = conditional.get('constraint')
            if isinstance(constraint, Mapping):
                constraint_mapping = cast(Mapping[str, object], constraint)
                constraint_type = constraint_mapping.get('type')
                if ('type' in constraint_mapping and constraint_type not in {'start_date', 'time_of_date'}) or not (
                    {'start_date', 'start_time', 'end_time'} & constraint_mapping.keys()
                ):
                    self._warn('constraint', f'{price_path}.constraint', context)
                    continue
            conditional['prices'] = self._project_prices(conditional['prices'], f'{price_path}.prices', context)
            prices.append(conditional)
        return prices

    def _project_price_map(self, raw: Mapping[str, object], path: str, context: str) -> dict[str, object]:
        prices: dict[str, object] = {}
        for price_key, value in raw.items():
            if isinstance(value, Mapping) and not ({'base', 'tiers'} & value.keys()):
                self._warn('price', f'{path}.{price_key}', context)
                continue
            prices[price_key] = value
        return prices

    def _warn(self, capability: str, path: str, context: str) -> None:
        self.warnings.append(
            f'Unsupported {capability} variant at {path} for {context}; upgrade genai-prices for full support'
        )


def _project_providers(raw: list[object]) -> tuple[list[object], tuple[str, ...]]:
    projector = _ProviderProjector()
    providers = [projector.project_provider(provider, index) for index, provider in enumerate(raw)]
    return providers, tuple(projector.warnings)


def _provider_context(provider: Mapping[str, object], index: int) -> str:
    provider_id = provider.get('id')
    return f'provider {provider_id!r}' if isinstance(provider_id, str) else f'provider index {index}'


def _model_context(model: Mapping[str, object], index: int) -> str:
    model_id = model.get('id')
    return f'model {model_id!r}' if isinstance(model_id, str) else f'model index {index}'
